File: Heaven-Engine/src/window/window.cpp
#include <window/window.hpp>

#include <iostream>

// static vars for tracking input

static bool isDebug = false;
static bool pressed = false;
static bool controllerCheck = false;

// callback function to move the OpenGL viewport to the GLFW window's position
static void framebuffer_size_callback(GLFWwindow* window, int width, int height){
    glViewport(0, 0, width, height);
}

// constructor, intializes GLFW and GLAD then creates a window with the passed parameters
Window::Window(int h, int w, const char* name) : App_State(ACTIVE), Input_State(KM), width(0), height(0){

    // set local vars of the window size
    width = w;
    height = h;

    // init GLFW
    glfwInit();
    glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 4);
    glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 5);
    glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE);
    glfwWindowHint(GLFW_RESIZABLE, false); //disable resizing the screen

    // create the window and check for errors
    handle = glfwCreateWindow(h, w, name, NULL, NULL);
    if(!handle){
        std::cout << "Failed to create window!" << std::endl;
        glfwTerminate();
        exit(-1);
    }

    // create window
    glfwMakeContextCurrent((GLFWwindow*)handle);

    // load glad
    if(!gladLoadGLLoader((GLADloadproc)glfwGetProcAddress)){
        std::cout << "Failed to init GLAD!" << std::endl;
        exit(-1);
    }

    // set up call back to update the opengl window
    glfwSetFramebufferSizeCallback((GLFWwindow*)handle, framebuffer_size_callback);
    // enable vsync
    glfwSwapInterval(1);
    // set openGL window size
    glViewport(0, 0, h, w);

    // set up rendering for 2D
    glEnable(GL_BLEND);
    glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);

    //TODO: Create debug options for the window class to display a console to show any errors or messages
    //std::cout << "MSG: Window successfully created" << std::endl;
}

// destructor
Window::~Window(){
    // delete any pointers
    glfwTerminate();
}

void Window::getInput(){

    // debug enabler button - toggle
    if(glfwGetKey((GLFWwindow*)handle, GLFW_KEY_GRAVE_ACCENT) == GLFW_PRESS && !pressed){
        isDebug = !isDebug;
        // check the isDebug value and set the proper app state
        if(isDebug){
            App_State = DEBUG;
            std::cout << "MSG: DEBUG IS ENABLED!" << std::endl;
        }
        else{
            App_State = ACTIVE;
            std::cout << "MSG: DEBUG IS DISABLED!" << std::endl;
        }
        pressed = !pressed;
    }else if(glfwGetKey((GLFWwindow*)handle, GLFW_KEY_GRAVE_ACCENT) == GLFW_RELEASE && pressed){
        pressed = !pressed;
    }

    //TODO: this checking system for joystick may need to check if joystick is indeed a gamepad
    // check for the first joystick
    if(glfwJoystickPresent(GLFW_JOYSTICK_1) == GLFW_TRUE && !controllerCheck){
        // print out that the joystick is connected
        const char* name = glfwGetJoystickName(GLFW_JOYSTICK_1);
        std::cout << "MSG: Controller is connected! ID: " << name << std::endl;
        std::cout << "MSG: Controller input is enabled!" << std::endl;
        Input_State = KMANDCONTROLLER;
        controllerCheck = !controllerCheck;
    }else if(glfwJoystickPresent(GLFW_JOYSTICK_1) == GLFW_FALSE && controllerCheck){
        std::cout << "MSG: Controller is disconnected!" << std::endl;
        std::cout << "MSG: Controller input is disabled!" << std::endl;
        Input_State = KM;
        controllerCheck = !controllerCheck;
    }

    // call additional input
    this->input();
}


float Window::getDeltaTime(){
    // calculate delta time
    currentFrame = glfwGetTime();
    this->DeltaTime = currentFrame - lastFrame;
    if(this->DeltaTime > 0.25)
        this->DeltaTime = 0.25;
    lastFrame = currentFrame;

    return this->DeltaTime;
}

// single threaded runtime of update(), stepUpdate() and render()
void Window::runtime(){
    while(!glfwWindowShouldClose((GLFWwindow*)this->handle)){
        // get Deltatime
        this->getDeltaTime();

        // check for glfw events
        glfwPollEvents();

        // check for main window input
        getInput();

        //  accumulate time and do stepUpdate()
        this->accumulator += this->DeltaTime;
        while(this->accumulator >= fixedTimeStep){
            // update with fixed time step
            stepUpdate(this->fixedTimeStep);
            accumulator -= fixedTimeStep;
        } 
        // calculate alpha for linear interpolation
        this->alpha = accumulator / this->fixedTimeStep;

        // update any input, values, objects, loading etc..
        update();

        // render background
        glClear(GL_COLOR_BUFFER_BIT);

        // draw or render
        render(alpha);

        // swap buffers
        glfwSwapBuffers((GLFWwindow*)this->handle);
    }
}

// handle main window input function
void Window::input(){
    // here goes additional input that is within SDL event handling and getEvent() loop
    // can be overwritten 
}

// initialization
void Window::init(){
    // here goes the initial processing of shaders, textues, and objects
}

//updating
void Window::update(){
    // update any variables like moving objects or updating input
    //* NOTE: that any object that needs input will need to have reference to the window handles a parameter to be passed down
}

// update physics or ticks
void Window::stepUpdate(double ts){
    // used to update physics, ticks, or other with a fixed time step 
}

// rendering
void Window::render(double alpha){
    // here update visually the objects, shaders, textures, etc
}

File: heaven.engine/src/main/java/engine/Window.java
package engine;

import org.lwjgl.*;
import org.lwjgl.glfw.*;
import org.lwjgl.opengl.*;
import org.lwjgl.system.*;

import java.util.Objects;

import static org.lwjgl.glfw.Callbacks.*;
import static org.lwjgl.glfw.GLFW.*;
import static org.lwjgl.opengl.GL45.*;
import static org.lwjgl.system.MemoryUtil.*;

public abstract class Window {

    private double last_frame;
    private double current_frame;
    private double accumulator;
    private int width;
    private int height;
    private double fixed_timeStep = 1.0f/60.0f;
    private double deltaTime;
    private long windowHandle;

    public Window(int w, int h, String name){
        // store the passed width and height
        width = w;
        height = h;
        // set up the rest of the variables to a default value
        last_frame = 0;
        current_frame = 0;
        accumulator = 0;
        deltaTime = 0;

        // initialize the window
        initializeWindow(name);
    }

    public double getDeltaTime(){
        current_frame = glfwGetTime();
        deltaTime = current_frame - last_frame;
        if(deltaTime > 0.25){
            deltaTime = 0.25;
        }
        last_frame = current_frame;
        return deltaTime;
    }

    private void initializeWindow(String name){
        // setup the GLFW error callback
        // by default print error messages in System.err
        GLFWErrorCallback.createPrint(System.err).set();

        // initialize glfw
        if(!glfwInit())
            throw new IllegalStateException("ERROR: Unable to initialize GLFW");

        // configure GLFW
        glfwDefaultWindowHints(); // the current window hints are already the defaults
        glfwWindowHint(GLFW_VISIBLE, GLFW_TRUE);
        glfwWindowHint(GLFW_RESIZABLE, GLFW_FALSE);
        glfwWindowHint(GLFW_MAXIMIZED, GLFW_FALSE);

        // create the window context and store into the "windowHandle" attribute
        windowHandle = glfwCreateWindow(width, height, name, NULL, NULL);
        // check the window handle
        if(windowHandle == NULL)
            throw new RuntimeException("ERROR: Failed to create the GLFW window");

        // make the OpenGL context
        glfwMakeContextCurrent(windowHandle);

        // create the OpenGL capabilities
        GL.createCapabilities();

        // set up OpenGL for 2D rendering
        glEnable(GL_BLEND);
        glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
    }

    public abstract void init();

    public abstract void update();

    public abstract void stepUpdate();

    public abstract void render();

    public void runtime(){
        // main loop of the window
        while(!glfwWindowShouldClose(windowHandle)){
            // get the deltaTime
            getDeltaTime();

            // check for any glfw events
            glfwPollEvents();

            // accumulate time and do stepUpdate()
            accumulator += deltaTime;
            while(accumulator >= fixed_timeStep){
                stepUpdate();
                accumulator -= fixed_timeStep;
            }

            //update any  values, objects, loading, etc..
            update();

            // clear the color buffer
            glClear(GL_COLOR_BUFFER_BIT);

            // draw or render
            render();

            // swap buffers
            glfwSwapBuffers(windowHandle);
        }
        // exit the window
        cleanUp();
    }

    private void cleanUp(){
        // Free the window callbacks and destroy the window
        glfwFreeCallbacks(windowHandle);
        glfwDestroyWindow(windowHandle);

        // Terminate GLFW and free the error callback
        glfwTerminate();
        Objects.requireNonNull(glfwSetErrorCallback(null)).free();
    }

    //unit test for window

}
